api/src/adapters/sam_gov/client.py

```diff
@@ -0,0 +1,242 @@
+
+        Returns:
+            Metadata about the downloaded file.
+
+        Raises:
+            ValueError: If the request parameters are invalid or the extract is not found.
+            IOError: If there is an error saving the file.
+            requests.exceptions.RequestException: If there is an error with the API request.
+        """
+        pass
+
+
+class SamGovClient(BaseSamGovClient):
+    """Client for interacting with the SAM.gov API."""
+
+    def __init__(
+        self,
+        api_key: str | SamGovConfig | None = None,
+        api_url: str | None = None,
+    ):
+        """Initialize the client.
+
+        Args:
+            api_key: API key for SAM.gov API or a SamGovConfig object.
+            api_url: URL for the SAM.gov API. If not provided, the value from environment variable will be used.
+        """
+        # Handle config object case
+        if isinstance(api_key, SamGovConfig):
+            config = api_key
+            self.api_key = config.api_key
+            self.api_url = config.base_url
+            self.extract_url = config.extract_url
+        else:
+            # Handle direct parameters or environment variables
+            config = SamGovConfig()
+            self.api_key = api_key or config.api_key
+            self.api_url = api_url or config.base_url
+            self.extract_url = config.extract_url
+
+        # Validate required parameters
+        if not self.api_url:
+            raise ValueError(
+                "API URL must be provided either directly, via config, or via environment variable."
+            )
+
+    def _build_headers(self) -> dict[str, str]:
+        """Build headers for API requests.
+
+        Returns:
+            Dictionary of headers.
+        """
+        return {
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        }
+
+    def download_extract(
+        self, extract_request: SamExtractRequest, output_path: str
+    ) -> SamExtractResponse:
+        """Download SAM.gov data extract.
+
+        Args:
+            extract_request: The request containing parameters for the extract download.
+                             Primarily used for the file_name to download.
+            output_path: Path to save the downloaded file. Can be a local file path or an S3 URI (s3://).
+
+        Returns:
+            Response object with download information.
+
+        Raises:
+            ValueError: If the request parameters are invalid or the extract is not found.
+            IOError: If there is an issue saving the file.
+            Exception: For any other errors.
+        """
+        # Validate that file_name is provided
+        if not extract_request.file_name:
+            raise ValueError("file_name must be provided for downloads")
+
+        # Validate that API key is provided
+        if not self.api_key:
+            raise ValueError("API key must be provided for SAM.gov API access")
+
```

Comment:
I was wrong about the api_key before - completely missed the small detail it had about being able to pass it as a header param. Completely missed the `The API Key value must be sent as "x-api-key" under "Headers" and not directly in the request URL.` alternative approach. Could we put it in the header? That feels like the most secure approach of the ones offered.